File: services/sync-base/app/utilities/s3connector.ts
import { randomFoo } from "../utilities/random";

export interface S3ConnectorConfig {
    batchSize?: number;
}

export class S3Connector<T> {
    private batchSize: number;

    constructor(config: S3ConnectorConfig) {
        this.batchSize = config.batchSize || 1000;
    }

    public async extract(): Promise<T[]> {
        console.log("Extracting data from S3");
        const data: T[] = [];
        for (let i = 0; i < this.batchSize; i++) {
            data.push(randomFoo() as T);
        }
        return data;
    }
}

File: services/connectors/src/s3connector.py
from typing import List, TypeVar, Generic, Optional
from .random import random_foo

T = TypeVar('T')

class S3ConnectorConfig:
    def __init__(self, batch_size: Optional[int] = None):
        self.batch_size = batch_size

class S3Connector(Generic[T]):
    def __init__(self, config: S3ConnectorConfig):
        self._batch_size = config.batch_size or 1000

    def extract(self) -> List[T]:
        print("Extracting data from S3")
        data: List[T] = []
        for i in range(self._batch_size):
            data.append(random_foo())
        return data